src/engine/strat_engine/dmdevice.rs

```diff
@@ -0,0 +1,142 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Functions for dealing with device mapper devices.
+
+use std::fmt;
+use std::fmt::Display;
+
+use super::super::super::engine::{EngineResult, EngineError, ErrorEnum, PoolUuid};
+
+const FORMAT_VERSION: u16 = 1;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Layer {
+    Flex,
+    Thin,
+    ThinPool,
+}
+
+impl Display for Layer {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Layer::Flex => write!(f, "flex"),
+            Layer::Thin => write!(f, "thin"),
+            Layer::ThinPool => write!(f, "thinpool"),
+        }
+    }
+}
+
+pub enum FlexRole {
+    MetadataVolume,
+    ThinData,
+    ThinMeta,
+}
+
+impl Display for FlexRole {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            FlexRole::MetadataVolume => write!(f, "mdv"),
+            FlexRole::ThinData => write!(f, "thindata"),
+            FlexRole::ThinMeta => write!(f, "thinmeta"),
+        }
+    }
+}
+
+pub enum ThinRole {
+    Filesystem,
+}
+
+impl Display for ThinRole {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ThinRole::Filesystem => write!(f, "fs"),
+        }
+    }
+}
+
+pub enum ThinPoolRole {
+    Pool,
+}
+
+impl Display for ThinPoolRole {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ThinPoolRole::Pool => write!(f, "pool"),
+        }
+    }
+}
+
+pub enum Role {
+    Flex(FlexRole),
+    Thin(ThinRole),
+    ThinPool(ThinPoolRole),
+}
+
+impl Display for Role {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Role::Flex(ref fr) => write!(f, "{}", fr),
+            Role::Thin(ref tr) => write!(f, "{}", tr),
+            Role::ThinPool(ref tr) => write!(f, "{}", tr),
+        }
+    }
+}
+
+/// Format a name for the flex layer.
+fn format_flex_name(pool_uuid: &PoolUuid, role: FlexRole) -> String {
+    return format!("stratis-{}-{}-flex-{}",
+                   FORMAT_VERSION,
+                   pool_uuid.simple().to_string(),
+                   role);
+}
+
+/// Format a name for the thin layer.
+fn format_thin_name(pool_uuid: &PoolUuid, role: ThinRole) -> String {
+    return format!("stratis-{}-{}-thin-{}",
+                   FORMAT_VERSION,
+                   pool_uuid.simple().to_string(),
+                   role);
+}
+
+/// Format a name for the thin pool layer.
+fn format_thinpool_name(pool_uuid: &PoolUuid, role: ThinPoolRole) -> String {
+    return format!("stratis-{}-{}-thinpool-{}",
+                   FORMAT_VERSION,
+                   pool_uuid.simple().to_string(),
+                   role);
+}
```

Comment:
The problem with that is that you loose certain benefits of type-checking. I've committed another approach.